File: LeetCode/Medium/0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        ret = []
        nums.sort()
        n = len(nums)-1

        for i in range(n-1):
            if i > 0 and nums[i] == nums[i-1]:
                continue
                
            l = i + 1
            r = n

            while l < r:
                temp = nums[l] + nums[r] + nums[i]
                if temp == 0:
                    ret.append([nums[l], nums[r], nums[i]])
                    l += 1
                    r -= 1
                    while l < r and nums[l] == nums[l-1]:
                        l += 1
                    while l < r and nums[r] == nums[r+1]:
                        r -= 1
                elif temp < 0:
                    l += 1
                else:
                    r -= 1

        return ret
                    
                    

File: LeetCode/Medium/0015-3sum/0015-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> ret = new ArrayList();
        int n = nums.length;
        
        for (int i = 0; i < n; ++i){
            if (i > 0 && nums[i] == nums[i-1]) continue;
            int l = i + 1;
            int r = n - 1;
            
            while (l < r){
                int sum = nums[i] + nums[l] + nums[r];
                if (sum == 0){
                    ret.add(Arrays.asList(nums[i], nums[l], nums[r]));
                    ++l;
                    while (l < r && nums[l] == nums[l-1]) ++l;
                    r--;
                    while (l < r && nums[r] == nums[r+1]) r--;
                }
                else if (sum < 0) ++l;
                else r--;
            }
        }
        
        return ret;
    }
}